src/core/cgroup.c

```diff
@@ -3863,6 +3866,21 @@ void unit_invalidate_cgroup_bpf(Unit *u) {
         }
 }
 
+void unit_cgroup_catchup(Unit *u) {
+        assert(u);
+
+        if (!UNIT_HAS_CGROUP_CONTEXT(u))
+                return;
+
+        /* We dropped the inotify watch during reexec/reload, so we need to
+         * check these as they may have changed.
+         * Note that (currently) the kernel doesn't actually update cgroup
+         * file modification times, so we can't just serialize and then check
+         * the mtime for file(s) we are interested in. */
+        (void) unit_check_cgroup_events(u);
+        (void) unit_check_oom(u);
```

Comment:
Calling `unit_check_oom()` bypasses the `cgroup_oom_queue` that the inotify watch would normally put the OOM-ed unit into. Shouldn't this at least check and remove the unit from `cgroup_oom_queue` so it keeps the list consistent?

I'm thinking of a situation where `cgroup_oom_queue` could have multiple items before `daemon-reload`, but afterwards the units would all be handled by `unit_check_oom()` without being removed from `cgroup_oom_queue`.

Reply:
AFAICS, `cgroup_oom_queue` isn't handled similarly to other queues, i.e. it's not gradually emptied from `manager_clear_jobs_and_units/unit_free`. (That looks like reload with a unit in `in_cgroup_oom_queue` can result in UAF throught the `m->cgroup_oom_queue`.)

So your comment is a relevant remark. I'd suggest handling the `cgroup_oom_queue` in accordance with other similar queues, i.e. empty it (w/out processing) before the reload and sync the unit's state anew after the reload.

Reply:
> Calling unit_check_oom() bypasses the cgroup_oom_queue

It's definitely strange that the `unit_check_oom`()/`unit_add_to_cgroup_oom_queue`() function pair is backwards from `unit_check_cgroup_events`()/`unit_add_to_cgroup_empty_queue`() function pair.

`unit_check_cgroup_events` actually does a check on `cgroup.events` and adds (or removes) the unit from the cgroup empty queue.

`unit_check_oom` however does the opposite; it expects to be called from `on_cgroup_oom_event` and does the actual handling of that event. This function seems misnamed to me.

I pushed a new patch that changes `unit_check_oom` to `unit_add_to_cgroup_oom_queue`, which seems correct, though the `unit_check_oom` and related functions could probably use some renaming/refactoring to be more consistent and clear.

Reply:
> I pushed a new patch that changes unit_check_oom to unit_add_to_cgroup_oom_queue, 

to clarify i didn't actually change those functions, only which one is now called from the `unit_cgroup_catchup` function